pq/observe_test.go

```diff
@@ -0,0 +1,392 @@
+				t.FatalOnError(err)
+			})
+			if t.Equal(statOnRead, stat.kind, "no read stat upon empty read") {
+				t.Equal(ReadStats{
+					Duration: stat.read.Duration,
+				}, stat.read)
+			}
+
+			qu.readWith(func(r *Reader) {
+				var tmp [7]byte
+				_, err := r.Read(tmp[:])
+				t.FatalOnError(err)
+			})
+			if t.Equal(statOnRead, stat.kind, "no read stat upon empty read") {
+				t.Equal(ReadStats{
+					Duration:   stat.read.Duration,
+					Read:       1,
+					BytesTotal: 9,
+					BytesMin:   9,
+					BytesMax:   9,
+				}, stat.read)
+			}
+		}))
+	})
+
+	t.Run("ack", func(t *mint.T) {
+		t.Run("fail on empty queue", withQueue(func(t *mint.T, qu *testQueue, stat *statEntry) {
+			err := qu.ACK(2)
+			t.Error(err)
+
+			t.Equal(statOnACK, stat.kind)
+			t.Equal(ACKStats{
+				Duration: stat.ack.Duration,
+				Failed:   true,
+				Events:   2,
+			}, stat.ack)
+		}))
+
+		t.Run("event in active page", withQueue(func(t *mint.T, qu *testQueue, stat *statEntry) {
+			qu.append("entry one", "entry two", "entry three", "entry four")
+			qu.flush()
+
+			t.FatalOnError(qu.ACK(1))
+			t.Equal(ACKStats{
+				Duration: stat.ack.Duration,
+				Failed:   false,
+				Events:   1,
+			}, stat.ack)
+		}))
+
+		t.Run("all events - single page", withQueue(func(t *mint.T, qu *testQueue, stat *statEntry) {
+			qu.append("entry one", "entry two", "entry three", "entry four")
+			qu.flush()
+
+			t.FatalOnError(qu.ACK(4))
+			t.Equal(ACKStats{
+				Duration: stat.ack.Duration,
+				Failed:   false,
+				Events:   4,
+				Pages:    0, // we don't free last page so to not interfere with concurrent write
+			}, stat.ack)
+		}))
+
+		t.Run("large events", withQueue(func(t *mint.T, qu *testQueue, stat *statEntry) {
+			qu.append(genEventLen(testPageSize*2.5), genEventLen(testPageSize*1.5))
+			qu.flush()
+
+			t.FatalOnError(qu.ACK(2))
+			t.Equal(ACKStats{
+				Duration: stat.ack.Duration,
+				Failed:   false,
+				Events:   2,
+				Pages:    2, // we only free pages for the first event, so to not mess with the readers local state
+			}, stat.ack)
+		}))
+	})
+}
+
+func newTestObserver(stat *statEntry) *testObserveLast {
+	return (*testObserveLast)(stat)
+}
+
+func (t *testObserveLast) OnQueueInit(off uintptr, version uint32, available uint) {
+	t.set(off, statOnOpen, statEntry{version: version, available: available})
+}
+
+func (t *testObserveLast) OnQueueFlush(off uintptr, stats FlushStats) {
+	t.set(off, statOnFlush, statEntry{flush: stats})
+}
+
+func (t *testObserveLast) OnQueueRead(off uintptr, stats ReadStats) {
+	t.set(off, statOnRead, statEntry{read: stats})
+}
+
+func (t *testObserveLast) OnQueueACK(off uintptr, stats ACKStats) {
+	t.set(off, statOnACK, statEntry{ack: stats})
+}
+
+func (t *testObserveLast) set(off uintptr, kind statKind, e statEntry) {
+	*t = testObserveLast(e)
+	t.off, t.kind = off, kind
+}
+
+func (s *statEntry) reset() { *s = statEntry{} }
+
+func genEventLen(n int) string {
+	return string(make([]byte, n))
+}
```

Comment:
@urso Seriously I have much to learn when looking at how you structure your tests, I like how you took something hard like a queue and were able to create simple assertions